pkg/kubelet/userns_manager.go

```diff
@@ -0,0 +1,600 @@
+	// fstore guarantees an atomic write, we need durability too.
+	parentDir, err := os.Open(dir)
+	if err != nil {
+		return err
+	}
+
+	if err = parentDir.Sync(); err != nil {
+		// Ignore return here, there is already an error reported.
+		parentDir.Close()
+		return err
+	}
+
+	return parentDir.Close()
+}
+
+// readMappingsFromFile reads the user namespace configuration from the pod directory.
+func (m *usernsManager) readMappingsFromFile(pod types.UID) ([]byte, error) {
+	dir := m.kl.getPodDir(pod)
+	fstore, err := utilstore.NewFileStore(dir, &utilfs.DefaultFs{})
+	if err != nil {
+		return nil, err
+	}
+	return fstore.Read(mappingsFile)
+}
+
+func MakeUserNsManager(kl userNsPodsManager) (*usernsManager, error) {
+	m := usernsManager{
+		// Create a bitArray for all the UID space (2^32).
+		// As a by product of that, no index param to bitArray can be out of bounds (index is uint32).
+		used:   makeBitArray((math.MaxUint32 + 1) / userNsLength),
+		usedBy: make(map[string]uint32),
+		kl:     kl,
+	}
+	// First block is reserved for the host.
+	m.used.set(0)
+
+	// Second block will be used for phase II. Don't assign that range for now.
+	m.used.set(1)
+
+	// do not bother reading the list of pods if user namespaces are not enabled.
+	if !utilfeature.DefaultFeatureGate.Enabled(features.UserNamespacesSupport) {
+		return &m, nil
+	}
+
+	found, err := kl.listPodsFromDisk()
+	if err != nil {
+		return nil, fmt.Errorf("user namespace manager can't read pods from disk: %w", err)
+	}
+	for _, podUID := range found {
+		klog.V(5).InfoS("reading pod from disk for user namespace", "podUID", podUID)
+		if err := m.recordPodMappings(podUID); err != nil {
+			return nil, err
+		}
+	}
+
+	return &m, nil
+}
+
+// recordPodMappings registers the range used for the user namespace if the
+// usernsConfFile exists in the pod directory.
+func (m *usernsManager) recordPodMappings(pod types.UID) error {
+	content, err := m.readMappingsFromFile(pod)
+	if err != nil && err != utilstore.ErrKeyNotFound {
+		return err
+	}
+	if string(content) == "" {
+		return nil
+	}
+
+	_, err = m.parseUserNsFileAndRecord(pod, content)
+	return err
+}
+
+// isSet checks if the specified index is already set.
+func (m *usernsManager) isSet(v uint32) bool {
+	index := v / userNsLength
+	return m.used.isSet(index)
+}
+
+// allocateOne finds a free user namespace and allocate it to the specified pod.
+// The first return value is the first ID in the user namespace, the second returns
+// the length for the user namespace range.
+func (m *usernsManager) allocateOne(pod string) (firstID uint32, length uint32, err error) {
+	if m.numAllocated >= maxPods {
+		return 0, 0, fmt.Errorf("limit on count of pods with user namespaces exceeded (limit is %v, current pods with userns: %v)", maxPods, m.numAllocated)
+	}
+	m.numAllocated++
+	defer func() {
+		if err != nil {
+			m.numAllocated--
+		}
+	}()
+
+	firstZero, found := m.used.findAvailable()
```

Comment:
isnt a mutex needed on this call between finding an available value and assigning it in the map?

Reply:
`GetUserNamespaceMappings()` already uses a lock when this is called:

`GetUserNamespaceMappings()` ->  `createUserNs()` -> `allocateOne()` -> `findAvailable()`